pygdf/columnops.py

```diff
@@ -153,33 +148,78 @@ def as_column(arbitrary):
 
     elif isinstance(arbitrary, np.ndarray):
         if arbitrary.dtype.kind == 'M':
-            # hack, coerce to int, then set the dtype
             data = datetime.DatetimeColumn.from_numpy(arbitrary)
         else:
-            data = as_column(Buffer(arbitrary))
-            if np.count_nonzero(np.isnan(arbitrary)) > 0:
-                mask = ~np.isnan(arbitrary)
-                data = data.set_mask(cudautils.compact_mask_bytes(mask))
-
-    elif isinstance(arbitrary, (list,)):
-        if any(x is None for x in arbitrary):
-            if all((isinstance(x, int) or x is None) for x in arbitrary):
-                padata = pa.array(arbitrary)
-                npdata = np.asarray(padata.buffers()[1]).view(np.int64)
-                data = as_column(npdata)
-                mask = np.asarray(padata.buffers()[0])
-                data = data.set_mask(mask)
-            elif all((isinstance(x, ) or x is None) for x in arbitrary):
-                padata = pa.array(arbitrary)
-                npdata = np.asarray(padata.buffers()[1]).view(np.float64)
-                data = as_column(npdata)
-                mask = np.asarray(padata.buffers()[0])
-                data = data.set_mask(mask)
+            data = as_column(cuda.to_device(arbitrary))
+
+    elif isinstance(arbitrary, pa.Array):
+        if isinstance(arbitrary, pa.StringArray):
+            raise NotImplementedError("Strings are not yet supported")
+        elif isinstance(arbitrary, pa.NullArray):
+            pamask = Buffer(np.empty(0, dtype='int8'))
+            padata = Buffer(
+                np.empty(0, dtype=arbitrary.type.to_pandas_dtype())
+            )
+            data = numerical.NumericalColumn(
+                data=padata,
+                mask=pamask,
+                null_count=0,
+                dtype=np.dtype(arbitrary.type.to_pandas_dtype())
+            )
+        elif isinstance(arbitrary, pa.DictionaryArray):
+            if arbitrary.buffers()[0]:
+                pamask = Buffer(np.array(arbitrary.buffers()[0]))
+            else:
+                pamask = None
+            padata = Buffer(np.array(arbitrary.buffers()[1]).view(
+                arbitrary.indices.type.to_pandas_dtype()
+            ))
+            data = categorical.CategoricalColumn(
+                data=padata,
+                mask=pamask,
+                null_count=arbitrary.null_count,
+                categories=arbitrary.dictionary.to_pylist(),
+                ordered=arbitrary.type.ordered,
+                dtype="category"  # What's the correct way to specify this?
```

Comment:
Shouldn't this be setup automatically since it is a CategoricalColumn or do we have alternative dtypes for these?

Reply:
You still have to pass a dtype from my testing. In our current `pandas_categorical_as_column` we pass the dtype from the input categorical object.